airflow/providers/common/sql/operators/sql.py

```diff
@@ -229,49 +237,81 @@ def __init__(
 
-            self.log.info("Record: %s", records)
+        self.log.info("Record: %s", records)
 
-            for idx, result in enumerate(records):
-                tolerance = self.column_mapping[column][checks[idx]].get("tolerance")
+        for row in records:
+            column, check, result = row
+            tolerance = self.column_mapping[column][check].get("tolerance")
 
-                self.column_mapping[column][checks[idx]]["result"] = result
-                self.column_mapping[column][checks[idx]]["success"] = self._get_match(
-                    self.column_mapping[column][checks[idx]], result, tolerance
-                )
+            self.column_mapping[column][check]["result"] = result
+            self.column_mapping[column][check]["success"] = self._get_match(
+                self.column_mapping[column][check], result, tolerance
+            )
 
-            failed_tests.extend(_get_failed_checks(self.column_mapping[column], column))
-        if failed_tests:
-            raise AirflowException(
-                f"Test failed.\nResults:\n{records!s}\n"
-                "The following tests have failed:"
-                f"\n{''.join(failed_tests)}"
+        for col, checks in self.column_mapping.items():
+            failed_tests.extend(
+                [
+                    f"Column: {col}\n\tCheck: {check},\n\tCheck Values: {check_values}\n"
+                    for check, check_values in checks.items()
+                    if not check_values["success"]
+                ]
             )
+        if failed_tests:
+            exception_string = f"""
+                Test failed.\nResults:\n{records!s}\n
+                The following tests have failed:
+                \n{''.join(failed_tests)}"""
+            _raise_exception(exception_string, self.retry_on_failure)
 
         self.log.info("All tests have passed")
 
+    def _generate_sql_query(self, column, checks):
+        def _generate_partition_clause(check):
+            if self.partition_clause and "partition_clause" not in checks[check]:
+                return f"WHERE {self.partition_clause}"
+            elif not self.partition_clause and "partition_clause" in checks[check]:
+                return f"WHERE {checks[check]['partition_clause']}"
+            elif self.partition_clause and "partition_clause" in checks[check]:
+                return f"WHERE {self.partition_clause} AND {checks[check]['partition_clause']}"
+            else:
+                return ""
+
+        checks_sql = "UNION ALL".join(
+            [
+                self.sql_check_template.format(
+                    check_statement=self.column_checks[check].format(column=column),
+                    check=check,
+                    table=self.table,
+                    column=column,
+                    partition_clause=_generate_partition_clause(check),
+                )
+                for check in checks
+            ]
+        )
```

Comment:
```suggestion
        checks_sql = "UNION ALL".join(
            self.sql_check_template.format(
                check_statement=self.column_checks[check].format(column=column),
                check=check,
                table=self.table,
                column=column,
                partition_clause=_generate_partition_clause(check),
            )
            for check in checks
        )
```

`join` takes a generator expression so the extra `[]` is not necessary (and they slow things down!)